src/foam/graphics/TreeNode.js

```diff
@@ -0,0 +1,297 @@
+/**
+ * @license
+ * Copyright 2019 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+  package: 'foam.graphics',
+  name: 'TreeNode',
+  extends: 'foam.graphics.Box',
+
+  requires: [
+    'foam.graphics.Box',
+    'foam.graphics.CView',
+    'foam.graphics.Label',
+    'foam.graphics.Line'
+  ],
+
+  imports: [
+    'formatNode',
+    'graph',
+    'nodeHeight',
+    'nodeWidth',
+    'padding',
+    'parentNode?',
+    'relationship'
+  ],
+  exports: ['as parentNode'],
+
+  properties: [
+    'data',
+    {
+      name: 'outline',
+      expression: function (x, nodeWidth, expanded, childNodes, padding) {
+        var nodeLeftEdgePadded = x - nodeWidth / 2 - padding / 2;
+        var nodeRightEdgePadded = x + nodeWidth / 2 + padding / 2;
+        var rootLevelOutline = [
+          {
+            left: nodeLeftEdgePadded,
+            right: nodeRightEdgePadded
+          }
+        ];
+
+        var champion = [];
+
+        for (let i = 0; i < childNodes.length && expanded; i++) {
+          // get child outline
```

Comment:
These comments are more "what" than "why". Comments that describe "what" is happening are often not useful and incur maintenance burden because they'll need to be updated when code is updated (which is often forgotten!) so I strongly recommend against doing this unless there's a good reason to do so. Explaining "why" the code is doing what it's doing is more useful because it provides context for why it was written in the first place.

Reply:
Made a better comment